## 2.1 Compute the number of days with positive and negative returns.

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv("TSLA.csv")
df_count = df.groupby(["Year"])["Return"].apply(lambda x: pd.Series([(x < 0).sum(), (x >= 0).sum()])).unstack()
df_count

0    1
Year          
2014  123  129
2015  126  126
2016  126  126
2017  117  134
2018  133  118
2019  120  132

#### According to the table above, there were 123 negative returns and 129 positive returns in 2014, 126 both negative and positive returns in 2015 and 2016, 117 negative returns and 134 positive returns in 2017, 133 negative and 118 positive returns in 2018, and 120 negative returns and 132 positive returns in 2019. 

## 2.2 Compute the average of daily returns 

In [2]:
df_m = pd.DataFrame(columns= ( "Trading days", "Mean", "% days < Mean", "% days >= Mean"))
df_m["Trading days"] = df.groupby(["Year"])["Year"].count()
df_m["Mean"] = df.groupby(["Year"])["Return"].mean()
less_mean = df.groupby(["Year"])["Return"].agg(lambda x: pd.Series([(x < x.mean()).sum()])).astype(int)
larger_mean = df.groupby(["Year"])["Return"].agg(lambda x: pd.Series([(x > x.mean()).sum()])).astype(int)
df_m["% days < Mean"] = less_mean / df_m["Trading days"]
df_m["% days >= Mean"] = larger_mean / df_m["Trading days"]
df_m.style.format({"% days < Mean": "{:.2%}", "% days >= Mean": "{:.2%}"})

#### Based on the table, the percentages of days above or below mean are around 50%.

## 2.3 Compute the number of days that returns are more than 2 standard deviations from the mean.

In [7]:
df_new = pd.DataFrame(columns= ( "Year","Return", "m", "2d"))
df_new["Year"] = df["Year"]
df_new["m"] = df.groupby(["Year"])["Return"].transform('mean')
df_new["2d"]= 2 * df.groupby(["Year"])["Return"].transform(np.std)
df_new["Return"] = df["Return"]
df_new["m-2d"] = df_new["m"] - df_new["2d"]
df_new["m+2d"] = df_new["m"] + df_new["2d"]
df_new["R < m-2d"] = df_new.apply(lambda x : "T" if x["Return"] < x["m-2d"] else "", axis=1)
df_new["R > m+2d"] = df_new.apply(lambda x : "T" if x["Return"] > x["m+2d"] else "", axis=1)

df_m_2d = pd.DataFrame(columns= ( "Trading days", "Mean(m)", "Standard Deviation(d)","% days < m-2d", "% days > m+2d"))
df_m_2d["Trading days"] = df_m["Trading days"]
df_m_2d["Mean(m)"] = df_m["Mean"]
df_m_2d["Standard Deviation(d)"] = df.groupby(["Year"])["Return"].std()
less_outlier = df_new.groupby(["Year"])["R < m-2d"].apply(lambda x: (x == "T").sum()).astype(int)
larger_outlier = df_new.groupby(["Year"])["R > m+2d"].apply(lambda x: (x == "T").sum()).astype(int)
df_m_2d["% days < m-2d"] = less_outlier / df_m_2d["Trading days"]
df_m_2d["% days > m+2d"] = larger_outlier / df_m_2d["Trading days"]
df_m_2d.style.format({"% days < m-2d": "{:.2%}", "% days > m+2d": "{:.2%}"})

    

#### The mean of daily returns are above 0 except the value in 2016, which implys that the overall trend of TSLA is positive and promising. For 2014, 2017 and 2018, the standard deviations are relatively higher than the other years, indicating that the daily returns are spread out over a larger range.